# STA130 Week 01 Homework 

Please see the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) for the list of topics covered in this homework assignment, and a list of topics that might appear during ChatBot conversations which are "out of scope" for the purposes of this homework assignment (and hence can be safely ignored if encountered)


<details class="details-example"><summary style="color:blue"><u>Introduction</u></summary>

### Introduction
    
A reasonable characterization of STA130 Homework is that it simply defines a weekly reading comprehension assignment. 
Indeed, STA130 Homework essentially boils down to completing various understanding confirmation exercises oriented around coding and writing tasks.
However, rather than reading a textbook, STA130 Homework is based on ChatBots so students can interactively follow up to clarify questions or confusion that they may still have regarding learning objective assignments.

> Communication is a fundamental skill underlying statistics and data science, so STA130 Homework based on ChatBots helps practice effective two-way communication as part of a "realistic" dialogue activity supporting underlying conceptual understanding building. 

It will likely become increasingly tempting to rely on ChatBots to "do the work for you". But when you find yourself frustrated with a ChatBots inability to give you the results you're looking for, this is a "hint" that you've become overreliant on the ChatBots. Your objective should not be to have ChatBots "do the work for you", but to use ChatBots to help you build your understanding so you can efficiently leverage ChatBots (and other resources) to help you work more efficiently.<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Instructions</u></summary>

### Instructions
    
1. Code and write all your answers (for both the "Prelecture" and "Postlecture" HW) in a python notebook (in code and markdown cells) 
    
> It is *suggested but not mandatory* that you complete the "Prelecture" HW prior to the Monday LEC since (a) all HW is due at the same time; but, (b) completing some of the HW early will mean better readiness for LEC and less of a "procrastentation cruch" towards the end of the week...
    
2. Paste summaries of your ChatBot sessions (including link(s) to chat log histories if you're using ChatGPT) within your notebook
    
> Create summaries of your ChatBot sessions by using concluding prompts such as "Please provide a summary of our exchanges here so I can submit them as a record of our interactions as part of a homework assignment" or, "Please provide me with the final working verson of the code that we created together"
    
3. Save your python jupyter notebook in your own account and "repo" on [github.com](github.com) and submit a link to that notebook though Quercus for assignment marking<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Prompt Engineering?</u></summary>

### Prompt Engineering?    
    
The questions (as copy-pasted prompts) are designed to initialize appropriate ChatBot conversations which can be explored in the manner of an interactive and dynamic textbook; but, it is nonetheless **strongly recommendated** that your rephrase the questions in a way that you find natural to ensure a clear understanding of the question. Given sensible prompts the represent a question well, the two primary challenges observed to arise from ChatBots are 

1. conversations going beyond the intended scope of the material addressed by the question; and, 
2. unrecoverable confusion as a result of sequential layers logial inquiry that cannot be resolved. 

In the case of the former (1), adding constraints specifying the limits of considerations of interest tends to be helpful; whereas, the latter (2) is often the result of initial prompting that leads to poor developments in navigating the material, which are likely just best resolve by a "hard reset" with a new initial approach to prompting.  Indeed, this is exactly the behavior [hardcoded into copilot](https://answers.microsoft.com/en-us/bing/forum/all/is-this-even-normal/0b6dcab3-7d6c-4373-8efe-d74158af3c00)...

</details>


### Marking Rubric (which may award partial credit) 

- [0.1 points]: All relevant ChatBot summaries [including link(s) to chat log histories if you're using ChatGPT] are reported within the notebook
- [0.2 points]: Reasonable well-written general definitions for Question "2.2"
- [0.3 points]: Demonstrated understanding regarding Question "4"
<!-- - [0.2 points]: A sensible justification for the choice in Question "7.4" -->
- [0.4 points]: Requested assessment of ChatBot versus google performance in Question "8.3"


## "Pre-lecture" HW [*completion prior to next LEC is suggested but not mandatory*]

#### 1. Pick one of the datasets from the ChatBot session(s) of the **TUT demo** (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If your TA has not shared a relevant ChatBot session from their **TUT demo** through a piazza post and a Quercus announcement, the **TUT notebook** has links to example ChatBot sessions that you can use; or, ...
> 
> ```python
> # feel free to just use the following if you prefer...
> import pandas as pd
> url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
> df = pd.read_csv(url)
> df.isna().sum()
> ```
    
</details>

In [2]:
import pandas as pd

# URL of the Squirrel Census dataset
url = "https://data.cityofnewyork.us/api/views/vfnx-vebw/rows.csv?accessType=DOWNLOAD"

# Read the dataset into a pandas DataFrame
df = pd.read_csv(url)

# Display the first few rows of the dataset
print(df.head())


           X          Y Unique Squirrel ID Hectare Shift      Date  \
0 -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1 -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2 -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3 -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4 -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   

   Hectare Squirrel Number    Age Primary Fur Color Highlight Fur Color  ...  \
0                        3    NaN               NaN                 NaN  ...   
1                        4    NaN               NaN                 NaN  ...   
2                        8    NaN              Gray                 NaN  ...   
3                       14  Adult              Gray                 NaN  ...   
4                        5  Adult              Gray            Cinnamon  ...   

    Kuks  Quaas  Moans Tail flags Tail twitches  Approaches  Indifferent  \
0  False  False  False

In [3]:
# Check for missing values in the dataset
missing_values = df.isnull().sum()

# Display columns with missing values
print(missing_values)


X                                                0
Y                                                0
Unique Squirrel ID                               0
Hectare                                          0
Shift                                            0
Date                                             0
Hectare Squirrel Number                          0
Age                                            121
Primary Fur Color                               55
Highlight Fur Color                           1086
Combination of Primary and Highlight Color       0
Color notes                                   2841
Location                                        64
Above Ground Sighter Measurement               114
Specific Location                             2547
Running                                          0
Chasing                                          0
Climbing                                         0
Eating                                           0
Foraging                       

# Chat log 

https://chatgpt.com/share/1c97311e-29c2-46cb-86ea-dbd0c5773bcd

#### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a `pandas` DataFrame has, and then

1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,  
2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> A good initial prompt to start would be be something like
> - "I've downloaded a dataset about characters from animal crossings (from https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv), and I'd like to know what columns of information I have and how much data I have"
> 
> You can further reduce the scope of your inquiry with if needed with something like
> - "I've already downloaded the data and want to understand the size (or dimensions) of the dataset to start with"
> 
> *Some ChatBots can upload your data and do this for you; but, extended usage of this feature [likely requires a paid subscription](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00006_gpt3p5_LoadDataPaywall.md); and, anyway, you need to run the code yourself rather than having a ChatBot do that for you; and, for STA130 we don't want a ChatBot to just do the analysis for us; rather, we instead want ChatBots to help us understand the steps we need to take to analyze the data; so,* **you DO NOT need to purchase an upgraded version of any ChatBots**
> - Free-tier level ChatBots like [GPT4o-mini](https://chat.openai.com/) or [Copilot](https://copilot.microsoft.com/) (which is partially based on [ChatGPT4.0](https://chat.openai.com/), and which you have access to through your UofT account) are sufficiently sophisticated and perfectly appropriate for the STA130 course
    
</details>

In [4]:
# Get the number of rows and columns
rows, columns = df.shape

print(f"The DataFrame has {rows} rows and {columns} columns.")

The DataFrame has 3023 rows and 31 columns.


Observations refers to the rows of the dataset, they are the individual entries of squirrel data (1 row = 1 squirrel)
Variables refers to the columns of the dataset, they are the attributes being used to describe each squirrel (e.g. age, primary fur colour)

#### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Use your ChatBot session to help you create working examples of using  `df.describe()` and `df['column'].value_counts()` for your dataset (although note that the `.value_counts()` method is not really meant to be used for numeric variables, so if you dataset has only numeric variables, `.value_counts()` might not be particularly informative...)
>
> #### ChatBot Response Scope 
>     
> If prompts are not sufficiently focused you will likely get overly broad responses from the ChatBot, but you can always respond with subsequent refinement requests to appropriately limit the scope of the ChatBot responses to focus on addressing your actual content targets; so, 
> - an initially very general inquiry like, "I need help analyzing my data" will likely result in a ChatBot response suggesting a wide variety of approaches and techniques for summarizing your dataset; but, re-prompting the ChatBot with something like, "What's the simplest form of summarization of this dataset that I could do and how do I do it in Python?" or suggesting guidance using the specific summarization methods requested above will helpfully re-orient the ChatBot to your specific interests and needs
> 
> #### Jupyter Notebook Hints
> 
> Jupyter notebook printouts usaully don't show all of the data (when there's too much to show, like if `df.describe()` includes results for many columns), but the printouts just show enough of the data to give an idea of what the results are which is all we're looking for at the moment
> 
> - Consider dividing the code that ChatBot provides you into different jupyter notebook cells so that each cell concludes with a key printed result; the last line of code in a jupyter notebook cell will automatically print out in a formatted manner, so replacing something like `print(df.head())` with `df.head()` as the last line of a cell provides a sensible way to organize your code
> - The printout suggestions above are demonstrated in `STA130F24_CourseProject.ipynb` if looking at an example would be helpful to understand what they're getting at...
    
</details>

In [5]:
pd.set_option('display.max_columns', None)

df.describe()

# can use df.describe(include = 'all') to show columns with non numerical values

X            Y          Date  Hectare Squirrel Number
count  3023.000000  3023.000000  3.023000e+03              3023.000000
mean    -73.967184    40.780853  1.011949e+07                 4.123718
std       0.007726     0.010285  4.246671e+04                 3.096492
min     -73.981159    40.764911  1.006202e+07                 1.000000
25%     -73.973102    40.771676  1.008202e+07                 2.000000
50%     -73.968594    40.778166  1.012202e+07                 3.000000
75%     -73.960189    40.791219  1.014202e+07                 6.000000
max     -73.949722    40.800119  1.020202e+07                23.000000

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3023 entries, 0 to 3022
Data columns (total 31 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   X                                           3023 non-null   float64
 1   Y                                           3023 non-null   float64
 2   Unique Squirrel ID                          3023 non-null   object 
 3   Hectare                                     3023 non-null   object 
 4   Shift                                       3023 non-null   object 
 5   Date                                        3023 non-null   int64  
 6   Hectare Squirrel Number                     3023 non-null   int64  
 7   Age                                         2902 non-null   object 
 8   Primary Fur Color                           2968 non-null   object 
 9   Highlight Fur Color                         1937 non-null   object 
 10  Combination 

In [7]:
# df.shape prints (rows, columns)
print(df.shape)

(3023, 31)


#### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by `df.shape` and what is reported by `df.describe()` with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If the dataset you're using does not have (a) non-numeric variables and (b) missing values in numeric variables (e.g., the `"villagers.csv"` example above has only a single numeric variable `row_n` which has no missing values), instead download and use the [https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv)" data to answer this question  
>
> In (a) above, the "columns it analyzes" refers to the columns of the output of `df.describe()` which will only include "numeric" columns by default, but you can can see the names of all the columns in a dataset using `df.columns`; and, make sure `df.shape` is refering to the dataset you think it is... if you've loaded a different dataset it might not have been called `df`(!)
>
> **If you get any errors (for example related to column names), copy and paste them as a response to the ChatBot, and see if it can help you resove them by adding the suggested adjustments to your code and then reruning all your code to see if the changes have fixed the problem (and repeat this process as needed until the problems have been resolved).**
    
</details>

### Q4 ANSWER

When a dataset has missing values, the number in the count column will exclude those values - and thus the count column will show a lower count than the df.shape command will show rows. If the data is complete, one would expect the row count in df.shape to match the 'count' column in df.describe. The number of columns analyzed in df.describe, and the number of columns displayed in df.shape are different, unless an extra parameter is added to df.describe to indicate that all columns should be shown. Otherwise, only numerical columns are displayed.

#### 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

- an "attribute", such as `df.shape` which does not end with `()`
- and a "method", such as `df.describe()` which does end with `()` 
   

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The fact that a "method" such as `df.describe()` ends with `()` suggests that "methods" are essentially something that we would call a "function" in programming language terminology; but, without getting too technical or "in the weeds", it might also be worth considering that we could also contrast what the difference is between a "function" in a programming language versus a "function" in mathematics...  
    
</details><br><br>

***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***<br><br>

### Q5 ANSWER

**Attributes** are properties of an object, there are inherent to the object and describe a characterstic of it.

**Methods** are actions that one can carry out on (or can be carried out by) an object

# Chat History

https://chatgpt.com/share/021011eb-ae65-4f73-a2d6-6dcd1498c532

# Chat Summary (from ChatGPT)
Certainly! Here’s a summary of our discussion:

1. Understanding DataFrame Dimensions
df.shape: Provides the dimensions of a DataFrame. It returns a tuple (number_of_rows, number_of_columns).
Usage Example:
python
Copy code
import pandas as pd
df = pd.read_csv('squirrel_census.csv')
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")
2. Observations vs. Variables
Observations (Rows): Individual data entries, each representing a single instance (e.g., a single squirrel sighting).
Variables (Columns): Different types of data or attributes recorded for each observation (e.g., location, date, color).
3. Summarizing Columns
df.describe(): Provides a statistical summary of numerical columns (mean, standard deviation, min, max, quartiles).
df.describe(include='object'): Provides summary statistics for categorical columns (counts, most frequent value).
df.describe(include='all'): Provides a summary for both numerical and categorical columns.
4. Displaying All Columns
Adjust Pandas Display Options: Use pd.set_option('display.max_columns', None) to display all columns.
Example:
python
Copy code
import pandas as pd
df = pd.read_csv('squirrel_census.csv')
pd.set_option('display.max_columns', None)
print(df.describe(include='all'))
pd.reset_option('display.max_columns')
5. Attributes vs. Methods
Attributes: Variables that represent the state or properties of an object (e.g., color, age).
Methods: Functions that define behaviors or actions an object can perform (e.g., make_sound(), describe()).
Attributes and methods help define and manipulate the state and behavior of objects in object-oriented programming.

Feel free to ask if you need more details or have additional questions!

<details class="details-example"><summary style="color:blue"><u>Continue now...?</u></summary>

### Prelecture VS Postlecture HW
    
Feel free to work on the "Postlecture" HW below if you're making good progress and want to continue: in this case this is particularly reasonable as questions "6" and "7" below directly follow up and extend the "Prelecture" HW questions

*The benefits of continue would are that (a) it might be fun to try to tackle the challenge of working through some problems without additional preparation or guidance; and (b) this is a very valable skill to be comfortable with; and (c) it will let you build experience interacting with ChatBots (and beginning to understand their strengths and limitations in this regard)... it's good to have sense of when using a ChatBot is the best way to figure something out, or if another approach (such as course provided resources or a plain old websearch for the right resourse) would be more effective*
    
</details>    


## "Post-lecture" HW [*submission along with "Pre-lecture" HW is due prior to next TUT*]

#### 6. The `df.describe()` method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The answers here actually make it obvious why these can only be calculated for numeric variables in a dataset, which should help explain the answer to "4(a)" and "4(b)" above
>   
> Also notice that when `df.describe()` is used missing values are not explicitly removed, but `df.describe()`  provides answers anyway. Is it clear what `df.describe()` does with the data in each columns it analyzes if there is missing data in the column in question? 
>
> The next questions addresses removing rows or columns from a dataset in order to explicitly remove the presense of any missingness in the dataset (assuming we're not going to fill in any missing data values using any missing data imputation methods, which are beyond the scope of STA130); so, the behavior of `df.describe()` hints that explicitly removing missing may not always be necessary; but, the concern, though, is that not all methods may be able to handle missing data the way `df.describe()` does...
    
</details>

**ANSWER** <br>
count: the number of non-null values in the column <br>
mean: the average of all the non-null values in the column <br>
std: the standard deviation of all the non-null values in the column <br>
min: the lowest value of all the non-null values in the column <br>
25%: quartile 1 (Q1), or the 25% percentile, the value at which 25% of the dataset is the same, or lower, numerically <br>
50%: the median, or the 50% percentile, the value at which 50% of the dataset is the same, or lower, numerically <br>
75%: quartile 3 (Q3), or the 75% percentile, the value at which 75% of the dataset is the same, or lower, numerically <br>
max: the highest value of all the non-null values in the column<br>

#### 7. Missing data can be considered "across rows" or "down columns".  Consider how `df.dropna()` or `del df['col']` should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

1. Provide an example of a "use case" in which using `df.dropna()` might be peferred over using `del df['col']`<br><br>
    
2. Provide an example of "the opposite use case" in which using `del df['col']` might be preferred over using `df.dropna()` <br><br>
    
3. Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important<br><br>
    
4. Remove all missing data from one of the datasets you're considering using some combination of `del df['col']` and/or `df.dropna()` and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.<br><br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Start a new ChatBot session **[but remember to first ask your ChatBot for summaries of your current session and perhaps coding results (so you can supply these in the homework as requested)]**, since your last ChatBot session has likely gotten quite long and has covered a lot of material at this point 
> - It can sometimes be helpful to reset ChatBot sessions to refocus them on the topics of inquiry without too much backlog history that might unintentionally bias things in certain directions and, of course, you can always re-introduce material from earlier conversations as it's relevant, such as for answering "D" based on reintroducing and updating code you made in a previous ChatBot session.  
> 
> #### ChatBot Scope Guidance
> 
> - This question is not interested in the general benefits of imputing missing data, or the general benefits of using `df.dropna()` and/or `del df['col']` to remove missing data, just how to most efficiently remove missing data if a user chooses to do so
> 
> - More sophisticated analyses for "filling in" rather than removing missing data (as considered here) are possible (based on making assumptions about missing data and using specific imputation methods or models) but these are "beyond the scope" of this homework assignment so this topics can be safely ignored for now
> 
> #### ChatBot Code Troubleshooting
> 
> A key issue to be aware of when asking ChatBots for help with something is that they are not running and checking code for correctess, and they often intertwine written instructions with code instructions; so, BEFORE YOU RUN ANY CODE provided by a ChatBot, you should check the following
> 
> 1. If this code changes an object or data, are you sure you want to run this code?
> 2. Can you easily "undo" the results of running code (e.g., from a copy `df_saved=df.copy()` or reloading the data) if running the code doesn't do what you want?
> 3. Is the state of the data what is expected by the code? Or have the objects been updated and changed so they're no longer what the code expects them to be? 
> 
> #### If you get any `Python` errors, copy and paste them into the ChatBot prompt and see if it can help you resove them; but, keep in mind the final point above becasue the ChatBot might not be aware of the state of your objects relative to the code it's producing...

</details><br>

1. del df can only be used on columns, and also permanently removes a column from the dataset. If the user wants to remove rows with missing data, or wants to keep the original dataset intact, df.dropna will be better. df.dropna has more flexibility than del df, which has only simple function. Also, df.dropna allows the user to automatically remove rows (or columns) with missing data - del df is not able to do that.

2. del df is typically faster, and simpler to execute. This indicates that del df will be more efficient, especially for large datasets where you already know which columns have null data that need to be removed. df.dropna can be less memory efficient as well, for the reasons stated above.

3. You may want to focus on specific columns to apply df.dropna() to, and so getting rid of unnecessary columns (perhaps with too many/entirely null values) first will make the usage of df.dropna() more efficient. There is also a lower risk of deleting important data, if you delete the columns with many null values first. The mostly-null column will likely be unimportant to your data, but if you apply df.dropna() first, then many important data points may be removed due to their having a null value in an irrelevant column.

In [8]:
# Many null values are present in some columns, some columns have only a few null values
# del df and df.dropna() will be used to remove all null values
# It is observed that there are 3023 data entries currently (pre-purge)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3023 entries, 0 to 3022
Data columns (total 31 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   X                                           3023 non-null   float64
 1   Y                                           3023 non-null   float64
 2   Unique Squirrel ID                          3023 non-null   object 
 3   Hectare                                     3023 non-null   object 
 4   Shift                                       3023 non-null   object 
 5   Date                                        3023 non-null   int64  
 6   Hectare Squirrel Number                     3023 non-null   int64  
 7   Age                                         2902 non-null   object 
 8   Primary Fur Color                           2968 non-null   object 
 9   Highlight Fur Color                         1937 non-null   object 
 10  Combination 

In [9]:
# Columns with less than 2872 non-null values were deleted 
# 95% of the column values being non-null was arbitrarily chosen as the cut-off
# Having 95% of the data values be non-null would mean around 2872 out of 3023 non-null values

del df['Color notes']
del df['Specific Location']
del df['Other Interactions']
del df['Other Activities']
del df['Highlight Fur Color']

In [10]:
# All the remaining rows with null values will be removed and added to a new dataframe

df_nonan = df.dropna()
df_nonan.info()

# It is observed that there are 2784 data entries currently (post-purge)

<class 'pandas.core.frame.DataFrame'>
Index: 2784 entries, 6 to 3022
Data columns (total 26 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   X                                           2784 non-null   float64
 1   Y                                           2784 non-null   float64
 2   Unique Squirrel ID                          2784 non-null   object 
 3   Hectare                                     2784 non-null   object 
 4   Shift                                       2784 non-null   object 
 5   Date                                        2784 non-null   int64  
 6   Hectare Squirrel Number                     2784 non-null   int64  
 7   Age                                         2784 non-null   object 
 8   Primary Fur Color                           2784 non-null   object 
 9   Combination of Primary and Highlight Color  2784 non-null   object 
 10  Location         

**Dataset Report** <br>
Before, there were 5 columns with a significant number (>95%) of null values. There were also some columns other columns with a small amount of null values. The del df function was first used to delete the 5 columns with null values, this reduced the risk of unnecessarily removing too many data entries when using df.dropna() later on. To delete the null values that remained after deleting the 5 columns, df.dropna() was used and stored in a new datafram, df_nonan. This new dataframe contained no null values. By using both df del and df.dropna() in combination, a minimal amount of data was lost in the process of obtaining this new dataframe. 

# Chat Link
https://chatgpt.com/share/78708869-5daa-4547-a166-6fc72d400b97


# Chat Summary

Summary of Our Conversation:
df.dropna() and del df['col']:

df.dropna(): Removes rows or columns containing null values, based on conditions such as axis=0 (rows) or axis=1 (columns). It's flexible and allows conditional removal based on null values.
del df['col']: Directly deletes a specific column from the DataFrame in place, without returning a new DataFrame or checking for conditions.
Differences Between drop() and del:

drop(): Allows for the conditional removal of rows/columns based on null values or other conditions. It offers flexibility and can return a new DataFrame or modify in place.
del: Is faster and simpler for removing a single column by name, but it only works for columns and is always in place.
Performance:

del is faster for directly deleting a column because it modifies the DataFrame in place without copying or evaluating conditions.
drop() can be slower due to condition evaluation and optional copying of the DataFrame.
Using drop() and del Together:

Combining these methods allows you to first remove unnecessary columns with del and then clean the remaining data with drop(). This approach optimizes both performance and accuracy in data cleaning.
Importance of Applying del Before dropna():

Performance: Deleting unnecessary columns first with del reduces the size of the DataFrame, making subsequent operations like dropna() faster and more efficient.
Accuracy: Deleting irrelevant columns before applying dropna() ensures that null values in unimportant columns don’t cause unintended data loss in the rows or columns you care about.
Logical Workflow: This sequence establishes a clear workflow where you first reduce irrelevant data and then clean the remaining data conditionally, preventing unnecessary row/column drops.





    
#### 8. Give brief explanations in your own words for any requested answers to the questions below

> This problem will guide you through exploring how to use a ChatBot to troubleshoot code using the "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" data set 
> 
> To initialially constrain the scope of the reponses from your ChatBot, start a new ChatBot session with the following slight variation on the initial prompting approach from "2" above
> - "I am going to do some initial simple summary analyses on the titanic data set I've downloaded (https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv) which has some missing values, and I'd like to get your help understanding the code I'm using and the analysis it's performing"
        
1. Use your ChatBot session to understand what `df.groupby("col1")["col2"].describe()` does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you
    
> If needed, you can help guide the ChatBot by showing it the code you've used to download the data **AND provide it with the names of the columns** using either a summary of the data with `df.describe()` or just `df.columns` as demonstrated [here](../CHATLOG/COP/00017_copilot_groupby.md)
    
2. Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the `count` value for different data columns depending on the missingness present in the original data.  Why do these capture something fundamentally different from the values in the `count` that result from doing something like `df.groupby("col1")["col2"].describe()`?

> Questions "4" and "6" above address how missing values are handled by `df.describe()` (which is reflected in the `count` output of this method); but, `count` in conjunction with `group_by` has another primary function that's more important than addressing missing values (although missing data could still play a role here).

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT<br><br>
    
    1. Forget to include `import pandas as pd` in your code 
       <br> 
       Use Kernel->Restart from the notebook menu to restart the jupyter notebook session unload imported libraries and start over so you can create this error
       <br><br>
       When python has an error, it sometimes provides a lot of "stack trace" output, but that's not usually very important for troubleshooting. For this problem for example, all you need to share with ChatGPT or search on google is `"NameError: name 'pd' is not defined"`<br><br>

    2. Mistype "titanic.csv" as "titanics.csv"
       <br> 
       If ChatBot troubleshooting is based on downloading the file, just replace the whole url with "titanics.csv" and try to troubleshoot the subsequent `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'` (assuming the file is indeed not present)
       <br><br>
       Explore introducing typos into a couple other parts of the url and note the slightly different errors this produces<br><br>
      
    3. Try to use a dataframe before it's been assigned into the variable
       <br> 
       You can simulate this by just misnaming the variable. For example, if you should write `df.groupby("col1")["col2"].describe()` based on how you loaded the data, then instead write `DF.groupby("col1")["col2"].describe()`
       <br><br>
       Make sure you've fixed your file name so that's not the error any more<br><br>
        
    4. Forget one of the parentheses somewhere the code
       <br>
       For example, if the code should be `pd.read_csv(url)` the change it to `pd.read_csv(url`<br><br>
        
    5. Mistype one of the names of the chained functions with the code 
       <br>
       For example, try something like `df.group_by("col1")["col2"].describe()` and `df.groupby("col1")["col2"].describle()`<br><br>
        
    6. Use a column name that's not in your data for the `groupby` and column selection 
       <br>
       For example, try capitalizing the columns for example replacing "sex" with "Sex" in `titanic_df.groupby("sex")["age"].describe()`, and then instead introducing the same error of "age"<br><br>
        
    7. Forget to put the column name as a string in quotes for the `groupby` and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question
       <br>
       For example, something like `titanic_df.groupby(sex)["age"].describe()`, and then `titanic_df.groupby("sex")[age].describe()`
        


## \#1, code and explanation

In [2]:
import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df2.describe()

df2.groupby('survived')['age'].describe()

count       mean        std   min   25%   50%   75%   max
survived                                                           
0         424.0  30.626179  14.172110  1.00  21.0  28.0  39.0  74.0
1         290.0  28.343690  14.950952  0.42  19.0  28.0  36.0  80.0

In [3]:
df2.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

df.groupby("col1")["col2"].describe() takes 2 columns as input. Column 1 is used as the independent variable, and column 2 is used as the dependent variable. Count, mean, std... of column 2 (age) are calculated for the categories in column 1 (survived or did not survive). Essentially, we are given values for: the count of people who survived, and didn't survive; the mean age of those who survived, and the mean age of those who didn't survive; the standard deviation of those who survived, and the standard deviation of those who didn't, and so on.

## \#2

Using df.describe() normally, there is a total count of 891 in the survived column. Presumably, this is the number of individuals in the entire dataset. However, the column for age only hasa count of 714. Thus, when using df.groupby('survived')\['age'\].describe(), the incomplete count must be used. Null values cannot be used in the calculations like standard deviation of age based on survival, and so null values are not considered when doing calculations using data from the col2.

## \#3

In [1]:
# forget import statement
# import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

# NameError: name 'pd' is not defined

NameError: name 'pd' is not defined

In [7]:
# file does not exist
import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv")

# HTTPError: HTTP Error 404: Not Found

# removing h from https -> ValueError: Protocol not known: ttps
# removing a forward slash from https:// -> URLError: <urlopen error no host given>
# removing : from https:// -> FileNotFoundError: [Errno 2] No such file or directory: 'https//raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'


In [8]:
# wrong dataset name

import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

DF2.describe()

# NameError: name 'DF2' is not defined

NameError: name 'DF2' is not defined

In [9]:
# missing parentheses

import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df2.describe(
    
# SyntaxError: incomplete input

SyntaxError: incomplete input (1724776906.py, line 7)

In [10]:
# typed function wrong

import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df2.describllee()

# AttributeError: 'DataFrame' object has no attribute 'describllee'

AttributeError: 'DataFrame' object has no attribute 'describllee'

In [14]:
# using data the doesn't exist

import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df2['AGE'].describe()

# KeyError: 'AGE'

KeyError: 'AGE'

In [17]:
# forgetting quotes for a column title

import pandas as pd

df2 = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df2[age].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: age, dtype: float64

# Chatbot vs Google for Errors

The chatbot was able to successfully identify the cause and solution for each error I gave it. I had to read through multiple possible options first, for some of the errors, since I wasn't able to give it much context, but it still gave me the right answer somewhere in its response. Google took longer to find the solution, since a lot of the given webpages were Quora or Stack Overflow posts of other people asking similar, but not applicable, questions. Google also told me everything about the error, but it wasn't specific enough to my purposes (errors relating to dataframes). 

It should be noted that the chatbot probably would've done even better if I had 1 continuous conversation with it. ChatGPT forced me to start a new conversation, so the entire new conversation was solely about errors with no context of the dataset. 

I would probably use ChatGPT for any errors I encounter in the future, rather than Google, unless I want a significantly deeper understanding of the error. Chatbots are excellent troubleshooters.

# Chat Logs and Links

https://chatgpt.com/share/528f4ff5-474b-413a-99c7-d5c34f4cf859 (1)

https://chatgpt.com/share/edc8893d-a2dc-49ac-ba84-fe8fc6a8b8fc (2)



# Summary 1

In this conversation, you asked for help with analyzing the Titanic dataset from a URL. I provided code to load and explore the dataset using pandas, including handling missing values and calculating summary statistics. You then inquired about the df.groupby("col1")["col2"].describe() function, and I explained that it provides descriptive statistics for each group in "col1" based on the values in "col2", excluding null values.

You encountered errors during the process, including:

A NameError due to pandas not being imported (resolved by adding import pandas as pd).
A 404 error for a missing dataset (potentially related to the URL or a different file).
A protocol typo (ttps instead of https) that caused a ValueError (resolved by correcting the URL to start with https://).
Now everything should work smoothly.

# Summary 2
Here's a summary of our conversation so far:

Error Handling: You provided several error messages related to Python, including:

NameError: name 'DF2' is not defined
SyntaxError: incomplete input
AttributeError: 'DataFrame' object has no attribute 'describllee'
KeyError: 'AGE'
NameError: name 'age' is not defined
Resolutions:

I explained that the NameError: name 'DF2' is not defined likely occurred because the variable DF2 wasn’t properly defined or initialized.
The SyntaxError: incomplete input suggested that there was incomplete code, possibly due to unclosed parentheses, braces, or blocks.
The AttributeError: 'DataFrame' object has no attribute 'describllee' was due to a typo, and I recommended using the correct describe() method.
The KeyError: 'AGE' pointed to trying to access a DataFrame column that doesn’t exist or is incorrectly named, and I suggested checking for typos or whitespace.
The NameError: name 'age' is not defined occurred because the variable age had not been defined, and I explained possible scope and case-sensitivity issues.
Feel free to provide more details if you'd like further clarification!

#### 9. Have you reviewed the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?<br>
    
<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Just answering "Yes" or "No" or "Somewhat" or "Mostly" or whatever here is fine as this question isn't a part of the rubric; but, the midterm and final exams may ask questions that are based on the tutorial and lecture materials; and, your own skills will be limited by your familiarity with these materials (which will determine your ability to actually do actual things effectively with these skills... like the course project...)
    
</details>
    
***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***

Yes!

# Recommended Additional Useful Activities [Optional]

The "Ethical Profesionalism Considerations" and "Current Course Project Capability Level" sections below **are not a part of the required homework assignment**; rather, they are regular weekly guides covering (a) relevant considerations regarding professional and ethical conduct, and (b) the analysis steps for the STA130 course project that are feasible at the current stage of the course<br><br>

<details class="details-example"><summary style="color:blue"><u>Ethical Professionalism Considerations</u></summary>

## Ethical Professionalism Considerations

> If the observed data is "no events occured" does this mean the data is "missing" and [should be ignored](https://priceonomics.com/the-space-shuttle-challenger-explosion-and-the-o)?
> 
> - NASA: \<determines temperature doesn't affects "o-ring" by subseting data to just "o-ring" incidents\>
> - Also NASA: \<launches the shuttle on a cold day\>

|No apparent "o-ring" failure and temperature relationship|Apparent between "o-ring" failure and temperature relationship|
|:-|:-|
if you just look at "o-ring" failure event data|if you instead look at ALL the data as you should|
|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image06-14.png)|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image02-33.png)|
|![](https://upload.wikimedia.org/wikipedia/commons/8/8b/Shuttle_Challenger_explosion.gif?20190203170223)|![](https://i.makeagif.com/media/10-04-2014/nT57xW.gif)|

<br>
    
</details>    

<details class="details-example"><summary style="color:blue"><u>Current Course Project Capability Level</u></summary>

## Current Course Project Capability Level

> The data we'll use for the STA130 course project is based on the [Canadian Social Connection Survey](https://casch.org/cscs). Please see the [data use agreement](https://static1.squarespace.com/static/60283c2e174c122f8ebe0f39/t/6239c284d610f76fed5a2e69/1647952517436/Data+Use+Agreement+for+the+Canadian+Social+Connection+Survey.pdf) regarding the appropriate and ethical professional use of this data (available at the bottom of the [CSCS](https://casch.org/cscs) webpage).
> 
> 1. Have a very quick look at the list of available variables available using the [link](https://drive.google.com/file/d/1ISVymGn-WR1lcRs4psIym2N3or5onNBi/view) (again at the bottom of the [CSCS](https://casch.org/cscs) webpage); then, 
> 2. examine the code in the first thirteen code cells of [STA130F24_CourseProject.ipynb](https://github.com/pointOfive/stat130chat130/blob/main/CP/STA130F24_CourseProject.ipynb) to get an initital understanding of how we might subset to different studies included in the [data](https://drive.google.com/file/d/1mbUQlMTrNYA7Ly5eImVRBn16Ehy9Lggo/view) (again accessible at the bottom of the [CSCS](https://casch.org/cscs) webpage); then,     
> 3. review the fourteenth and fifteenth cells (with the comments "Here's a high level summary of the data" and "And here are some explanations about the columns in the data") a little more closely to get a better sense of which columns seem to be the most interesting and whether or not they seem to have a lot of missing data
    
</details>        

### Afterward

Here are few ideas of some other kinds of interactions you might consider exploring with a ChatBot...

> While these are likely to be extremely practically valuable, they are not a part of the homework assignment, so do not include anything related to these in your homework submission

- With respect to improving ones ability in statistics, coding, communication, and other key data science skills
    - what is the ChatBots perception its own capabilities and uses as an AI-driven assistance tool 
    - and does ChatBots assessment of itself influence or agree with your own evalution of the ChatBot? 

- ChatBots can introduce and explain the "World War 2 planes" problem and the "Monte Hall" problem... 
    - how well does do they seem to do and introducing and explaining other "unintuitive surprising statistics paradoxes"?

- If you consider the process of writing about why you chose to take this course, and the skills you were hoping to build through this course with respect to your current ideas about what possible careers 
    - and how do you think the exercise would be different if you framed it as a dialogue with a ChatBot
    - and do you think the difference could be positive and productive, or potentially biasing and distracting?
    
- ChatBots sometimes immediately responds in simple helpful ways, but other times it gives a lot of extraneous information that can be overwheling... are you able to prompt and interact with ChatBots in manner that keeps its reponses helpful and focused on what you're interested in? 

- ChatBots tends to respond in a fairly empathetic and supportive tone...
    - do you find it helpful to discuss concerns you might have about succeeding in the course (or entering university more generally) with a ChatBot?
    
- For what purposes and in what contexts do you think a ChatBot could provide suggestions or feedback about your experiences that might be useful? 
